File: JavaScript solutions/350. Intersection of Two Arrays II.js
//approach : 1
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {
  nums1.sort((a, b) => a - b);
  nums2.sort((a, b) => a - b);
  let ans = [];
  var i = 0,
    j = 0;
  while (i < nums1.length && j < nums2.length) {
    if (nums1[i] === nums2[j]) {
      ans.push(nums1[i]);
      i++;
      j++;
    } else if (nums1[i] < nums2[j]) i++;
    else j++;
  }
  return ans;
};

//approach : 2
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {
  const map = new Map();
  for (const num of nums1) {
    map.set(num, map.get(num) + 1 || 1);
  }
  const ans = [];
  for (const num of nums2) {
    if (map.has(num) && map.get(num) > 0) {
      ans.push(num);
      map.set(num, map.get(num) - 1);
    }
  }
  return ans;
};

//approach : 3
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {
  let ans = [];
  for (var i = 0; i < nums1.length; i++) {
    for (var j = 0; j < nums2.length; j++) {
      if (nums1[i] === nums2[j]) {
        ans.push(nums1[i]);
        nums2[j] = -1;
        break;
      }
    }
  }
  return ans;
};

File: C++ solutions/350. Intersection of Two Arrays II.cpp
#include<bits/stdc++.h>
using namespace std;

// solution
//approach : 1
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(),nums1.end());
        sort(nums2.begin(),nums2.end());
        vector<int> ans;
        int i=0,j=0;
        while(i<nums1.size() && j<nums2.size())
        {
            if(nums1[i]==nums2[j])
            {
                ans.push_back(nums1[i]);
                i++;
                j++;
            }
            else if(nums1[i]<nums2[j]) i++;
            else j++;
        }
        return ans;
    }
};

//approach : 2
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        map<int,int> count;
        vector<int> ans;
        for(int i=0;i<nums1.size();i++)
        {
            count[nums1[i]]++;
        }
        
        for(int i=0;i<nums2.size();i++)
        {
            if(count[nums2[i]])
            {
                ans.push_back(nums2[i]);
                count[nums2[i]]--;
            }
        }
        return ans;
    }
};

//approach : 3
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int>v;
        for(int i=0;i<nums1.size();i++){
            for(int j=0;j<nums2.size();j++){
                if(nums1[i]==nums2[j]){
                    v.push_back(nums1[i]);
                    nums2[j]=-1;
                    break;
                }
            }
        }
        return v;
    }
};

// approach : 4
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        map<int,int>m1,m2;
        vector<int>v;
        for(int i=0;i<nums1.size();i++){
            m1[nums1[i]]++;
        }
        for(int i=0;i<nums2.size();i++){
            m2[nums2[i]]++;
        }
        for(auto i:m1){
            for(auto j:m2){
                if(i.first==j.first){
                    for(int k=0;k<min(i.second,j.second);k++) v.push_back(i.first);
                }
            }
        }
        return v;
    }
};